### PR TITLE
Improve async functions control

### PR DESCRIPTION
**Pull request by RedMickey**

### CODE REVIEW

File: src/cdp-proxy/debuggerEndpointHelper.ts [hunk 1]
```diff
@@ -23,17 +26,29 @@ export class DebuggerEndpointHelper {
...
             return await this.getWSEndpoint(browserURL);
         } catch (e) {
-            if (attemptNumber < 1) {
-                throw new Error(`Could not connect to debug target at ${browserURL}: ${e.message}`);
-              }
+            if (attemptNumber < 1 || cancellationToken.isCancellationRequested) {
+                const error = ErrorHelper.getInternalError(InternalErrorCode.CouldNotConnectToDebugTarget, browserURL, e.message);
+
+                if (cancellationToken.isCancellationRequested) {
+                    const nestedError = ErrorHelper.getNestedError(error, InternalErrorCode.CancellationTokenTriggered);
```
Comment (SounD120) [inline src/cdp-proxy/debuggerEndpointHelper.ts:43]:
```suggestion
                    throw ErrorHelper.getNestedError(error, InternalErrorCode.CancellationTokenTriggered);
```

File: src/cdp-proxy/debuggerEndpointHelper.ts [hunk 2]
```diff
@@ -23,17 +26,29 @@ export class DebuggerEndpointHelper {
...
+                    throw nestedError;
```
Comment (SounD120) [inline src/cdp-proxy/debuggerEndpointHelper.ts:44]:
```suggestion
```

File: src/cdp-proxy/debuggerEndpointHelper.ts [hunk 3]
```diff
@@ -23,17 +26,29 @@ export class DebuggerEndpointHelper {
...
+        cancellationToken: CancellationToken
+    ): Promise<string> {
         try {
             return await this.getWSEndpoint(browserURL);
         } catch (e) {
-            if (attemptNumber < 1) {
-                throw new Error(`Could not connect to debug target at ${browserURL}: ${e.message}`);
-              }
+            if (attemptNumber < 1 || cancellationToken.isCancellationRequested) {
+                const error = ErrorHelper.getInternalError(InternalErrorCode.CouldNotConnectToDebugTarget, browserURL, e.message);
```
Comment (SounD120) [inline src/cdp-proxy/debuggerEndpointHelper.ts:40]:
```suggestion
```

File: src/cdp-proxy/debuggerEndpointHelper.ts [hunk 4]
```diff
@@ -23,17 +26,29 @@ export class DebuggerEndpointHelper {
...
+
+                if (cancellationToken.isCancellationRequested) {
+                    const nestedError = ErrorHelper.getNestedError(error, InternalErrorCode.CancellationTokenTriggered);
+                    throw nestedError;
+                }
+
+                throw error;
```
Comment (SounD120) [inline src/cdp-proxy/debuggerEndpointHelper.ts:47]:
```suggestion
                throw ErrorHelper.getInternalError(InternalErrorCode.CouldNotConnectToDebugTarget, browserURL, e.message);
```

File: src/debugger/direct/directDebugSession.ts [hunk 5]
```diff
@@ -93,37 +93,27 @@ export class DirectDebugSession extends DebugSessionBase {
                             logger.log(`Connecting to ${attachArgs.port} port`);
                             return this.appLauncher.getRnCdpProxy().stopServer()
                                 .then(() => this.appLauncher.getRnCdpProxy().initializeServer(new DirectCDPMessageHandler(), this.cdpProxyLogLevel))
-                                .then(() => this.debuggerEndpointHelper.retryGetWSEndpoint(`http://localhost:${attachArgs.port}`, 90))
+                                .then(() => this.debuggerEndpointHelper.retryGetWSEndpoint(
+                                    `http://localhost:${attachArgs.port}`,
+                                    90,
+                                    this.cancellationTokenSource.token
```
Comment (SounD120) [inline src/debugger/direct/directDebugSession.ts:99]: What if do several attaches simultaneously? Will this token be the same for all these attaches, or new session and token will be generated?

File: src/debugger/debugSessionBase.ts [hunk 6]
```diff
@@ -104,10 +117,34 @@ export abstract class DebugSessionBase extends LoggingDebugSession {
...
+                logger.warn(localize("CouldNotStopMonitoringLogcat", "Couldn't stop monitoring logcat: {0}", err.message || err));
+            }
+        }
+
+        super.disconnectRequest(response, args, request);
+    }
+
+    protected showError(error: Error, response: DebugProtocol.Response): void {
+        if ((error instanceof InternalError || error instanceof NestedError)
+            && error.errorCode === InternalErrorCode.CancellationTokenTriggered
```
Comment (SounD120) [inline src/debugger/debugSessionBase.ts:140]: Do we need to completely stop showing errors if session is stopped? We may miss some important information for debugging some bugs if we do this. What do you think? Or maybe I am missing something?

File: src/debugger/debugSessionBase.ts [hunk 7]
```diff
@@ -48,21 +50,32 @@ export interface ILaunchRequestArgs extends DebugProtocol.LaunchRequestArguments
...
         this.isSettingsInitialized = false;
-        this.appWorker = null;
         this.debugSessionStatus = DebugSessionStatus.FirstConnection;
+        this.cancellationTokenSource = new vscode.CancellationTokenSource();
+    }
+
+    protected initializeRequest(response: DebugProtocol.InitializeResponse, args: DebugProtocol.InitializeRequestArguments): void {
+        response.body = response.body || {};
+
+        response.body.supportsConfigurationDoneRequest = true;
```
[the remark this answers is not in the document]
Comment (RedMickey) [reply src/debugger/debugSessionBase.ts:73]: Yes, all these options are documented on the official DAP site. You can find more info [here](https://microsoft.github.io/debug-adapter-protocol/specification#Types_Capabilities).

File: src/cdp-proxy/debuggerEndpointHelper.ts [hunk 8]
```diff
@@ -23,17 +26,28 @@ export class DebuggerEndpointHelper {
      * Attempts to retrieve the debugger websocket URL for a process listening
      * at the given address, retrying until available.
      * @param browserURL -- Address like `http://localhost:1234`
+     * @param cancellationToken -- Optional cancellation for this operation
```
Comment (SounD120) [inline src/cdp-proxy/debuggerEndpointHelper.ts:29]: It is documented like optional, but it is used as a required argument.

**Pull request merged**